sbws/core/scanner.py

```diff
@@ -131,7 +132,25 @@ def measure_bandwidth_to_server(session, conf, dest, content_length):
     return results
 
 
+def obtain_ed25519(controller, relay):
+    """Obtain ed25519 identity of the relay. Modifies relay object.
+
+    :param stem.control.Controller controller: stem Tor _init_controller
+    :param Relay: the relay
+    """
+    if relay.identifier is None:
+        try:
+            log.debug('Getting microdescriptor to obtain ed25519 identity.')
+            mdesc = controller.get_microdescriptor(relay.fingerprint)
```

Comment:
This code only works for clients with "UseMicrodescriptors 1" (or auto), which is the default.

But many torflow clients have "UseMicrodescriptors 0", and we want to support clients that have server descriptors. So if getting the microdescriptor fails, please fall back to getting the server descriptor:

https://stem.torproject.org/api/control.html#stem.control.Controller.get_server_descriptor

Reply:
You'll want the ed25519_master_key of the RelayDescriptor:
https://stem.torproject.org/api/descriptor/server_descriptor.html#stem.descriptor.server_descriptor.RelayDescriptor